app/models/provider/twelve_data.rb

```diff
@@ -0,0 +1,193 @@
+  Error = Class.new(Provider::Error)
+  InvalidExchangeRateError = Class.new(Error)
+  InvalidSecurityPriceError = Class.new(Error)
+
+  def initialize(api_key)
+    @api_key = api_key
+  end
+
+  def healthy?
+    with_provider_response do
+      response = client.get("#{base_url}/api_usage")
+      JSON.parse(response.body).dig("plan_category").present?
+    end
+  end
+
+  def usage
+    with_provider_response do
+      response = client.get("#{base_url}/api_usage")
+
+      parsed = JSON.parse(response.body)
+
+      limit = parsed.dig("plan_daily_limit")
+      used = parsed.dig("daily_usage")
+      remaining = limit - used
+
+      UsageData.new(
+        used: used,
+        limit: limit,
+        utilization: used / limit * 100,
+        plan: parsed.dig("plan_category"),
+      )
+    end
+  end
+
+  # ================================
+  #          Exchange Rates
+  # ================================
+
+  def fetch_exchange_rate(from:, to:, date:)
+    with_provider_response do
+      response = client.get("#{base_url}/exchange_rate?symbol=#{from}/#{to}") do |req|
+        req.params["date"] = date.to_s
+      end
+
+      rate = JSON.parse(response.body).dig("rate")
+
+      Rate.new(date: date.to_date, from:, to:, rate: rate)
+    end
+  end
+
+  def fetch_exchange_rates(from:, to:, start_date:, end_date:)
+    with_provider_response do
+      request_body = {}
+      date = start_date
+      while date <= end_date
+        request_body[date.to_s] = { url: "/exchange_rate?symbol=#{from}/#{to}&date=#{date.to_s}" }
+        date += 1.day
+      end
+
+      response = client.post("#{base_url}/batch", request_body)
+      data = JSON.parse(response.body).dig("data")
+      data.map do |date, resp|
+        rate = resp.dig("response", "rate")
+        if rate.nil?
+          Rails.logger.warn("#{self.class.name} returned invalid rate data for pair from: #{from} to: #{to} on: #{date}.  Rate data: #{rate.inspect}")
+          next
+        end
+
+        Rate.new(date: date.to_date, from:, to:, rate:)
+      end.compact
+    end
+  end
+
+  # ================================
+  #           Securities
+  # ================================
+
+  def search_securities(symbol, country_code: nil, exchange_operating_mic: nil)
+    with_provider_response do
+      response = client.get("#{base_url}/symbol_search") do |req|
+        req.params["symbol"] = symbol
+        req.params["outputsize"] = 25
+      end
+
+      parsed = JSON.parse(response.body)
+
+      parsed.dig("data").map do |security|
+        Security.new(
+          symbol: security.dig("symbol"),
+          name: security.dig("instrument_name"),
+          logo_url: "https://placehold.co/40?text=#{security.dig("symbol")}", # TODO: Twelve Data does not provide logo URLs
+          exchange_operating_mic: security.dig("mic_code"),
+          country_code: "US" # TODO: country_code is not provided by Twelve Data, so we default to US
```

Comment:
Twelve provides the full country name instead of the country code, what would eb the best way to do a conversion here?

Reply:
Added a countries gem to handle this conversion